### PR TITLE
Add bitmask feature in cx command 

### PR DESCRIPTION
**Pull request by kamou**
Hello,
I find it useful to be able to compare signature like hex patterns.
I suggest the "cs" command for this purpose (see commit)

One could use it as follows:
```
[0x08048e78]> px 4
- offset -   0 1  2 3  4 5  6 7  8 9  A B  C D  E F  0123456789ABCDEF
0x08048e78  52ba 5c8a                                R.\.            
[0x08048e78]> cs 52..5c
Compare 3/3 equal bytes (100%)
```

I plan to add '*' as byte matching wildcard. Would you take it ?
If you have any better idea on the command or its implementation, please let me know.

Ayman

### CODE REVIEW

**Comment by radare** [comment]
Whats the difference with cx or otherc subcommands?

**Comment by kamou** [comment]
`cs 52..5c` would match everything from 52005c to 53FF5c.
I might have missed a command doing the same ?

**Comment by radare** [comment]
Why 53?

**Comment by kamou** [comment]
typo !
read 52FF5c

**Comment by radare** [comment]
iirc this is what `cx` is suposed to do. so i think that this implementation shuold be under `cx` instead of adding a new one. cx right now doesnt honors the `.` bitmask.

also, can you add some tests for this?

**Comment by radare** [comment]
also, i think that having cx* for spitting out r2 commands instead of human readable stuff can be also interesting

**Comment by kamou** [comment]
Sure I can do that in cx instead.
I agree that I never really look at these human readable logs, just using `??` in my scripts.
Should it for example generate commands that seek to first failing comparison ?
Would you happen to have any other specific ideas about c* ?

File: libr/core/cmd_cmp.c [hunk 1]
```diff
@@ -367,16 +369,30 @@ static int cmd_cmp(void *data, const char *input) {
...
 		if (input[1]!=' ') {
-			eprintf ("Usage: cx 001122'\n");
+			eprintf ("Usage: cx 00..22'\n");
 			return 0;
 		}
+		filled = (char*) malloc(strlen(input+2)+1);
+		if (filled == NULL)
+			return false;
+
+		memcpy(filled, input + 2, strlen(input+2)+1);
```
Comment (radare) [inline libr/core/cmd_cmp.c]: missing spaces before every parenthesys and around +

File: libr/core/cmd_cmp.c [hunk 2]
```diff
@@ -367,16 +369,30 @@ static int cmd_cmp(void *data, const char *input) {
 		break;
 	case 'x':
 		if (input[1]!=' ') {
-			eprintf ("Usage: cx 001122'\n");
+			eprintf ("Usage: cx 00..22'\n");
 			return 0;
 		}
+		filled = (char*) malloc(strlen(input+2)+1);
```
Comment (radare) [inline libr/core/cmd_cmp.c]: missing spaces

File: libr/core/cmd_cmp.c [hunk 3]
```diff
@@ -367,16 +369,30 @@ static int cmd_cmp(void *data, const char *input) {
...
+			return false;
+
+		memcpy(filled, input + 2, strlen(input+2)+1);
+
 		buf = (ut8*)malloc (strlen (input+2)+1);
 		if (buf == NULL)
 			return false;
-		ret = r_hex_str2bin (input+2, buf);
+
+		ret = r_hex_bin2str (core->block, strlen(input+2)/2, (char *)buf);
```
Comment (radare) [inline libr/core/cmd_cmp.c]: missing spaces

**Comment by radare** [comment]
looks good. please fix those spacing issues and i'll merge

thanks!

**Comment by kamou** [comment]
done !
could you please let me know what r2 command cx* is expected to output ?

**Comment by radare** [comment]
Maybe something like:

wx 83 @ $$+4

Where 83 is the new byte and +4 the offset

**Comment by kamou** [comment]
ok I see.
would that mean that the new byte would be one of the parameters provided to cx* ?
something like:
```
wx 00112233
cx* 00112244 33
```
would output:
`wx 33 @ $$+3`

am I right ?
also as the user would like to check the offset and the value before, this would also be appreciated:
```
p8 1 @ $$+3
?e 3
wx 33 @ $$+3

```
First line would give the failing data
Second line the offset
And third the final patch the user could execute if the his checks on the first lines are ok.

**Comment by kamou** [comment]
oh I guess I missed the point..
maybe you meant:
```
wx 00112233
cx* 00112244
```
would output:
`wx 44 @ $$+3`

**Comment by radare** [comment]
yep!

**Comment by radare** [comment]
can you update the pr so i can merge?

**Comment by kamou** [comment]
It is already updated regarding the spacing issues.
I you are referring to cx*, I didn't get the time to do it yet.

**Comment by kamou** [comment]
now I added the * variants.

**Comment by radare** [comment]
Thanks!

**Pull request closed without merging**